src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2208 @@
+                                                "halide_xtensa_slice_to_native",
+                                                {args[arg_index], ix, native_lanes * slicing_multiplier, total_lanes},
+                                                Call::PureExtern);
+                    }
+                    sliced_args.push_back(sliced_arg);
+                }
+
+                Expr r = Call::make(op->type.with_lanes(native_lanes * slicing_multiplier), op->name, sliced_args, op->call_type);
+                concat_args.push_back(std::move(r));
+            }
+
+            return Call::make(op->type,
+                              "halide_xtensa_concat_from_native",
+                              concat_args, Call::PureExtern);
+        }
+
+        // TODO(vksnk): need to be careful here, because not everything can be
+        // padded safely.
+        int width_to_extend = get_width_to_extend(op->type);
+        bool is_safe_to_pad = true;
+        for (const auto &arg : op->args) {
+            is_safe_to_pad = is_safe_to_pad && (arg.type().is_scalar() || (op->type.lanes() == arg.type().lanes()));
+        }
+        std::set<std::string> safe_to_pad = {"halide_xtensa_dynamic_shuffle"};
+        is_safe_to_pad = is_safe_to_pad || (safe_to_pad.count(op->name) > 0);
+        std::set<std::string> skip_padding = {"halide_xtensa_widening_load"};
+        is_safe_to_pad = is_safe_to_pad && (skip_padding.count(op->name) == 0);
+        if (width_to_extend > 0 && is_safe_to_pad) {
+            vector<Expr> args;
+            const int lanes = op->type.lanes();
+
+            for (const auto &arg : op->args) {
+                Expr padded_arg;
+                if (arg.type().is_scalar()) {
+                    padded_arg = arg;
+                } else {
+                    Expr mutated_arg = mutate(arg);
+                    padded_arg = pad(mutated_arg, lanes, width_to_extend);
+                }
+
+                args.push_back(padded_arg);
+            }
+
+            Expr r = Call::make(op->type.with_lanes(width_to_extend), op->name, args, op->call_type);
+
+            return slice(r, op->type, lanes);
+        }
+
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const VectorReduce *op) override {
+        // TODO(vksnk): Factor it out.
+        Expr (*binop)(Expr, Expr) = nullptr;
+        switch (op->op) {
+        case VectorReduce::Add:
+            binop = Add::make;
+            break;
+        case VectorReduce::Mul:
+            binop = Mul::make;
+            break;
+        case VectorReduce::Min:
+            binop = Min::make;
+            break;
+        case VectorReduce::Max:
+            binop = Max::make;
+            break;
+        case VectorReduce::And:
+            binop = And::make;
+            break;
+        case VectorReduce::Or:
+            binop = Or::make;
+            break;
+        case VectorReduce::SaturatingAdd:
+            binop = saturating_add;
```

Comment:
Change this to Halide::saturating_add (the one in the internal namespace is the deprecated wrapper)